### ISSUE TITLE
[Module Proposal]: `avm-ptn-function-app-secured-storage`

### ISSUE THREAD

**Issue by kewalaka** [Azure/Azure-Verified-Modules#770]
### Check for previous/existing GitHub issues/module proposals

- [X] I have checked for previous/existing GitHub issues/module proposals.

### Check this module doesn't already exist in the module indexes

- [X] I have checked for that this module doesn't already exist in the module indexes; or I'm proposing the module to be migrated from CARML/TFVM.

### Bicep or Terraform?

Terraform

### Module Classification?

Pattern Module

### Module Name

avm-ptn-function-app-secured-storage

### Module Details

This is a pattern module proposal for a function app using secured storage & private endpoints.

i.e. it is an implementation of this quickstart template:

https://learn.microsoft.com/en-us/samples/azure/azure-quickstart-templates/function-app-storage-private-endpoints/

I have a start of this implemented here:
https://github.com/kewalaka/terraform-azurerm-avm-ptn-function-app-secured-storage

### Do you want to be the owner of this module?

No

### (Optional) Secondary Module Owner's GitHub Username (handle)

kewalaka

**Comment by kewalaka** [comment]
@donovm4 I'm using the AVM function app module that you're the owner of to build this one out.

Are you interested in being the owner of this too?

**Comment by donovm4** [comment]
@kewalaka - this looks interesting. i am cool with owning this.

just to ask, is there a specific time table for this pattern?

**Comment by kewalaka** [comment]
hi @donovm4 - i'm currently working on this pattern for a customer, so I expect to have something functional for my use case this week (there's a good start on the [feat/initial_version](https://github.com/kewalaka/terraform-azurerm-avm-ptn-function-app-storage-private-endpoints/tree/feat/initial_version) branch)

I'm happy to drive the work within the scope of my use case - which will be equivalent to the quick start capability, which I think makes a good v0.1.0 - after that it depends what interest there is to add improvements

**Comment by donovm4** [comment]
@kewalaka - gotcha! I am/will be available to assist however you need for v0.1.0

I do have some breaking changes to make for the avm-res-web-site module (currently on web_app_development branch) allowing for deployment of both function apps and web apps, and then will also make push the `managed_identites` fixes. I haven't pushed/published them yet, but we can communicate before I do, that way I don't harm your progress so far.

**Comment by prjelesi** [comment]
HI to both, based on above conversation:

@donovm4 

Thanks for proposing to be an AVM module owner!

We just want to confirm you agree to the below pages that define what module ownership means:

[Team Definitions & RACI](https://azure.github.io/Azure-Verified-Modules/specs/shared/team-definitions)
[Shared Specification (Bicep & Terraform)](https://azure.github.io/Azure-Verified-Modules/specs/shared)
[Module Support](https://azure.github.io/Azure-Verified-Modules/help-support/module-support)
Any questions or clarifications needed, let us know!

If you agree, please just reply to this issue with the exact sentence below (as this helps with our automation 👍):

"I CONFIRM I WISH TO OWN THIS AVM MODULE AND UNDERSTAND THE REQUIREMENTS AND DEFINITION OF A MODULE OWNER"

Thanks,

The AVM Core Team

**Comment by donovm4** [comment]
I CONFIRM I WISH TO OWN THIS AVM MODULE AND UNDERSTAND THE REQUIREMENTS AND DEFINITION OF A MODULE OWNER

**Comment by prjelesi** [comment]
@donovm4 pls confirm module name. 

avm-ptn-function-app-secured-storage
or
avm-ptn-function-app-storage-private-endpoints

?

**Comment by donovm4** [comment]
@prjelesi - let's continue with the second one. that is fine, as I trust your judgement

**Comment by kewalaka** [comment]
> i would go for second one to align with template name.
> 
> ?

Agreed that's why I renamed the repo, ..  thought it was good to align to the quickstart template name

**Comment by prjelesi** [comment]
Good!

@donovm4 , 

Thanks for confirming that you wish to own this AVM module and understand the related requirements and responsibilities!

Before starting development, please ensure ALL the following requirements are met.

Please use the following values explicitly as provided in the [module index](https://azure.github.io/Azure-Verified-Modules/indexes/) page:

For your module:
ModuleName - for naming your module
TelemetryIdPrefix - for your module's [telemetry](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-sfr3---category-telemetry---deploymentusage-telemetry)
For your module's repository:
Repo name and folder path are defined in RepoURL
Create GitHub teams for module owners and contributors and grant them permissions as outlined [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr20---category-contributionsupport---github-teams-only).
Grant permissions for the AVM core team and PG teams on your GitHub repo as described [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr9---category-contributionsupport---avm--pg-teams-github-repo-permissions).
Check if this module exists in the other IaC language. If so, collaborate with the other owner for consistency. 👍

You can now start the development of this module! 

@donovm4 , @kewalaka ✅ Happy coding! 🎉

Please respond to this comment and request a review from the AVM core team once your module is ready to be published! Please include a link pointing to your PR, once available. 🙏

Any further questions or clarifications needed, let us know!

Thanks,

The AVM Core Team

**Comment by donovm4** [comment]
@kewalaka - ended up getting a review and pushing some v0.2 changes for the function / web app module. 

Since you have that feat/initial branch, would you like me to start that branch and migration on the Azure repo?

**Comment by kewalaka** [comment]
@donovm4 i can merge that branch in to a dev branch if you like?  Is the Azure org repo publicly available?

it would be nice to be credited with the initial work in the commit history, helps when i'm selling the AVM concept :-)

it might be easier to ping me on email (on my profile) then I can contact you back via Teams or similar :-)

**Comment by donovm4** [comment]
@kewalaka - no problem at all! will reach out soon.

it is currently still going through review so it private at the moment

**Comment by diberry** [comment]
@donovm4 Can you update this issue? I'm look for an Azure Functions app module with no passwords or connection strings.

**Comment by donovm4** [comment]
hi @diberry : 

the module is not yet published, but the repo is located [here](https://github.com/Azure/terraform-azurerm-avm-ptn-function-app-storage-private-endpoints).

I had to shift focus to adding features to the app service resource module, and will be continuing development of this pattern soon.

Feel free to reach out through Teams / email / pattern repo with any feedback or questions 👍

**Comment by donovm4** [comment]
I have completed my initial development of the module and I would like to request a review of my module before publishing it to the terraform registry. The latest code is in a PR titled [AVM-Review-PR](https://github.com/Azure/terraform-azurerm-avm-ptn-function-app-storage-private-endpoints/pull/14) on the module repo and all checks on that PR are passing. *Kinda*

**Comment by donovm4** [comment]
Hi Team:

Just wanted to update this thread. I am still waiting on a review so that I can get this published. Is there anyone available over the next couple of weeks who can walk through a review?

CC: @Azure/avm-core-team-technical

**Comment by donovm4** [comment]
The initial review of this module is complete, and the module has been published to the registry. Requesting AVM team to close this module proposal and mark the module available in the module index.
Terraform Registry Link: [avm-ptn-function-app-storage-private-endpoints](https://registry.terraform.io/modules/Azure/avm-ptn-function-app-storage-private-endpoints/azurerm/latest)
GitHub Repo Link: [Azure/terraform-azurerm-avm-ptn-function-app-storage-private-endpoints](https://github.com/Azure/terraform-azurerm-avm-ptn-function-app-storage-private-endpoints)
